Backend Test Flake (test_creates_stats_model_from_one_suggestion_in_set_format): Datetime Mismatch

The backend test `core.jobs.batch_jobs.suggestion_stats_computation_jobs_test.GenerateTranslationContributionStatsJobTests.test_creates_stats_model_from_one_suggestion_in_set_format` flaked due to a mismatch of Datetime objects. This is likely because midnight UTC hit while the test was running.

Log:

```
======================================================================
FAIL: test_creates_stats_model_from_one_suggestion_in_set_format (core.jobs.batch_jobs.suggestion_stats_computation_jobs_test.GenerateTranslationContributionStatsJobTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/oppia/oppia/core/jobs/batch_jobs/suggestion_stats_computation_jobs_test.py", line 296, in test_creates_stats_model_from_one_suggestion_in_set_format
    self.assertItemsEqual(
  File "/home/runner/work/oppia/oppia/core/tests/test_utils.py", line 1745, in assertItemsEqual
    return super().assertCountEqual(*args, **kwargs)
AssertionError: Element counts were not equal:
First has 1, Second has 0:  datetime.date(2022, 11, 3)
First has 0, Second has 1:  datetime.date(2022, 11, 4)

Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/runner/work/oppia/oppia/scripts/run_backend_tests.py", line 613, in <module>
    main()
  File "/home/runner/work/oppia/oppia/scripts/run_backend_tests.py", line 530, in main
    raise Exception(
Exception: 0 errors, 1 failures
---------------------------------------------------------------------
```

Link: https://github.com/oppia/oppia/actions/runs/3390103589/jobs/5633868052

--- a/core/jobs/batch_jobs/suggestion_stats_computation_jobs_test.py
+++ b/core/jobs/batch_jobs/suggestion_stats_computation_jobs_test.py
@@ -1187,7 +1187,7 @@ class GenerateContributionStatsJobTests(job_test_utils.JobTestBase):
                 target_id=self.EXP_1_ID,
                 target_version_at_submission=0,
                 language_code=self.LANG_1,
-                created_on=datetime.datetime.utcnow()
+                created_on=mocked_now
             )
             suggestion_1_model.update_timestamps()
             suggestion_2_model = self.create_model(
@@ -1209,7 +1209,7 @@ class GenerateContributionStatsJobTests(job_test_utils.JobTestBase):
                 target_id=self.EXP_1_ID,
                 target_version_at_submission=0,
                 language_code=self.LANG_1,
-                created_on=datetime.datetime.utcnow() - datetime.timedelta(days=2)
+                created_on=mocked_now - datetime.timedelta(days=2)
             )
             suggestion_2_model.update_timestamps()
             suggestion_3_model = self.create_model(
@@ -1231,7 +1231,7 @@ class GenerateContributionStatsJobTests(job_test_utils.JobTestBase):
                 target_id=self.EXP_1_ID,
                 target_version_at_submission=0,
                 language_code=self.LANG_1,
-                created_on=datetime.datetime.utcnow() - datetime.timedelta(days=1)
+                created_on=mocked_now - datetime.timedelta(days=1)
             )
             suggestion_3_model.update_timestamps()
             suggestion_models.GeneralSuggestionModel.put_multi([

--- a/core/jobs/batch_jobs/suggestion_stats_computation_jobs_test.py
+++ b/core/jobs/batch_jobs/suggestion_stats_computation_jobs_test.py
@@ -77,40 +77,42 @@ class GenerateContributionStatsJobTests(job_test_utils.JobTestBase):
         self.assert_job_output_is_empty()
 
     def test_creates_stats_model_from_one_in_review_suggestion(self) -> None:
-        suggestion_model = self.create_model(
-            suggestion_models.GeneralSuggestionModel,
-            suggestion_type=feconf.SUGGESTION_TYPE_TRANSLATE_CONTENT,
-            author_id=self.VALID_USER_ID_1,
-            change_cmd={
-                'cmd': exp_domain.CMD_ADD_WRITTEN_TRANSLATION,
-                'state_name': 'state',
-                'content_id': 'content_id',
-                'language_code': 'lang',
-                'content_html': '111 222 333',
-                'translation_html': '111 222 333',
-                'data_format': 'html'
-            },
-            score_category='irelevant',
-            status=suggestion_models.STATUS_IN_REVIEW,
-            target_type='exploration',
-            target_id=self.EXP_1_ID,
-            target_version_at_submission=0,
-            language_code=self.LANG_1
-        )
-        suggestion_model.update_timestamps()
-        suggestion_model.put()
-        opportunity_model = self.create_model(
-            opportunity_models.ExplorationOpportunitySummaryModel,
-            id=self.EXP_1_ID,
-            topic_id=self.TOPIC_1_ID,
-            chapter_title='irelevant',
-            content_count=1,
-            story_id='irelevant',
-            story_title='irelevant',
-            topic_name='irelevant'
-        )
-        opportunity_model.update_timestamps()
-        opportunity_model.put()
+        mocked_now = datetime.datetime(2025, 1, 7)
+        with self.mock_datetime_utcnow(mocked_now):
+            suggestion_model = self.create_model(
+                suggestion_models.GeneralSuggestionModel,
+                suggestion_type=feconf.SUGGESTION_TYPE_TRANSLATE_CONTENT,
+                author_id=self.VALID_USER_ID_1,
+                change_cmd={
+                    'cmd': exp_domain.CMD_ADD_WRITTEN_TRANSLATION,
+                    'state_name': 'state',
+                    'content_id': 'content_id',
+                    'language_code': 'lang',
+                    'content_html': '111 222 333',
+                    'translation_html': '111 222 333',
+                    'data_format': 'html'
+                },
+                score_category='irelevant',
+                status=suggestion_models.STATUS_IN_REVIEW,
+                target_type='exploration',
+                target_id=self.EXP_1_ID,
+                target_version_at_submission=0,
+                language_code=self.LANG_1
+            )
+            suggestion_model.update_timestamps()
+            suggestion_model.put()
+            opportunity_model = self.create_model(
+                opportunity_models.ExplorationOpportunitySummaryModel,
+                id=self.EXP_1_ID,
+                topic_id=self.TOPIC_1_ID,
+                chapter_title='irelevant',
+                content_count=1,
+                story_id='irelevant',
+                story_title='irelevant',
+                topic_name='irelevant'
+            )
+            opportunity_model.update_timestamps()
+            opportunity_model.put()
 
         self.assert_job_output_is([
             job_run_result.JobRunResult(
@@ -198,39 +200,41 @@ class GenerateContributionStatsJobTests(job_test_utils.JobTestBase):
     def test_creates_stats_model_from_one_suggestion_in_legacy_format(
         self
     ) -> None:
-        suggestion_model = self.create_model(
-            suggestion_models.GeneralSuggestionModel,
-            suggestion_type=feconf.SUGGESTION_TYPE_TRANSLATE_CONTENT,
-            author_id=self.VALID_USER_ID_1,
-            change_cmd={
-                'cmd': exp_domain.DEPRECATED_CMD_ADD_TRANSLATION,
-                'state_name': 'state',
-                'content_id': 'content_id',
-                'language_code': 'lang',
-                'content_html': '111 a',
-                'translation_html': '111 a'
-            },
-            score_category='irelevant',
-            status=suggestion_models.STATUS_IN_REVIEW,
-            target_type='exploration',
-            target_id=self.EXP_1_ID,
-            target_version_at_submission=0,
-            language_code=self.LANG_1
-        )
-        suggestion_model.update_timestamps()
-        suggestion_model.put()
-        opportunity_model = self.create_model(
-            opportunity_models.ExplorationOpportunitySummaryModel,
-            id=self.EXP_1_ID,
-            topic_id=self.TOPIC_1_ID,
-            chapter_title='irelevant',
-            content_count=1,
-            story_id='irelevant',
-            story_title='irelevant',
-            topic_name='irelevant'
-        )
-        opportunity_model.update_timestamps()
-        opportunity_model.put()
+        mocked_now = datetime.datetime(2025, 1, 7)
+        with self.mock_datetime_utcnow(mocked_now):
+            suggestion_model = self.create_model(
+                suggestion_models.GeneralSuggestionModel,
+                suggestion_type=feconf.SUGGESTION_TYPE_TRANSLATE_CONTENT,
+                author_id=self.VALID_USER_ID_1,
+                change_cmd={
+                    'cmd': exp_domain.DEPRECATED_CMD_ADD_TRANSLATION,
+                    'state_name': 'state',
+                    'content_id': 'content_id',
+                    'language_code': 'lang',
+                    'content_html': '111 a',
+                    'translation_html': '111 a'
+                },
+                score_category='irelevant',
+                status=suggestion_models.STATUS_IN_REVIEW,
+                target_type='exploration',
+                target_id=self.EXP_1_ID,
+                target_version_at_submission=0,
+                language_code=self.LANG_1
+            )
+            suggestion_model.update_timestamps()
+            suggestion_model.put()
+            opportunity_model = self.create_model(
+                opportunity_models.ExplorationOpportunitySummaryModel,
+                id=self.EXP_1_ID,
+                topic_id=self.TOPIC_1_ID,
+                chapter_title='irelevant',
+                content_count=1,
+                story_id='irelevant',
+                story_title='irelevant',
+                topic_name='irelevant'
+            )
+            opportunity_model.update_timestamps()
+            opportunity_model.put()
 
         self.assert_job_output_is([
             job_run_result.JobRunResult(
@@ -272,40 +276,43 @@ class GenerateContributionStatsJobTests(job_test_utils.JobTestBase):
     def test_creates_stats_model_from_one_suggestion_in_set_format(
         self
     ) -> None:
-        suggestion_model = self.create_model(
-            suggestion_models.GeneralSuggestionModel,
-            suggestion_type=feconf.SUGGESTION_TYPE_TRANSLATE_CONTENT,
-            author_id=self.VALID_USER_ID_1,
-            change_cmd={
-                'cmd': exp_domain.CMD_ADD_WRITTEN_TRANSLATION,
-                'state_name': 'state',
-                'content_id': 'content_id',
-                'language_code': 'lang',
-                'content_html': ['111 a', '222 b', '333 c'],
-                'translation_html': ['111 a', '222 b', '333 c'],
-                'data_format': 'set_of_normalized_string'
-            },
-            score_category='irelevant',
-            status=suggestion_models.STATUS_IN_REVIEW,
-            target_type='exploration',
-            target_id=self.EXP_1_ID,
-            target_version_at_submission=0,
-            language_code=self.LANG_1
-        )
-        suggestion_model.update_timestamps()
-        suggestion_model.put()
-        opportunity_model = self.create_model(
-            opportunity_models.ExplorationOpportunitySummaryModel,
-            id=self.EXP_1_ID,
-            topic_id=self.TOPIC_1_ID,
-            chapter_title='irelevant',
-            content_count=1,
-            story_id='irelevant',
-            story_title='irelevant',
-            topic_name='irelevant'
-        )
-        opportunity_model.update_timestamps()
-        opportunity_model.put()
+         # Define a fixed datetime.
+        mocked_now = datetime.datetime(2025, 1, 7)
+        with self.mock_datetime_utcnow(mocked_now):
+            suggestion_model = self.create_model(
+                suggestion_models.GeneralSuggestionModel,
+                suggestion_type=feconf.SUGGESTION_TYPE_TRANSLATE_CONTENT,
+                author_id=self.VALID_USER_ID_1,
+                change_cmd={
+                    'cmd': exp_domain.CMD_ADD_WRITTEN_TRANSLATION,
+                    'state_name': 'state',
+                    'content_id': 'content_id',
+                    'language_code': 'lang',
+                    'content_html': ['111 a', '222 b', '333 c'],
+                    'translation_html': ['111 a', '222 b', '333 c'],
+                    'data_format': 'set_of_normalized_string'
+                },
+                score_category='irelevant',
+                status=suggestion_models.STATUS_IN_REVIEW,
+                target_type='exploration',
+                target_id=self.EXP_1_ID,
+                target_version_at_submission=0,
+                language_code=self.LANG_1
+            )
+            suggestion_model.update_timestamps()
+            suggestion_model.put()
+            opportunity_model = self.create_model(
+                opportunity_models.ExplorationOpportunitySummaryModel,
+                id=self.EXP_1_ID,
+                topic_id=self.TOPIC_1_ID,
+                chapter_title='irelevant',
+                content_count=1,
+                story_id='irelevant',
+                story_title='irelevant',
+                topic_name='irelevant'
+            )
+            opportunity_model.update_timestamps()
+            opportunity_model.put()
 
         self.assert_job_output_is([
             job_run_result.JobRunResult(
@@ -347,40 +354,42 @@ class GenerateContributionStatsJobTests(job_test_utils.JobTestBase):
     def test_creates_stats_model_from_one_in_review_suggestion_with_opportunity(
         self
     ) -> None:
-        suggestion_model = self.create_model(
-            suggestion_models.GeneralSuggestionModel,
-            suggestion_type=feconf.SUGGESTION_TYPE_TRANSLATE_CONTENT,
-            author_id=self.VALID_USER_ID_1,
-            change_cmd={
-                'cmd': exp_domain.CMD_ADD_WRITTEN_TRANSLATION,
-                'state_name': 'state',
-                'content_id': 'content_id',
-                'language_code': 'lang',
-                'content_html': '111 222 333',
-                'translation_html': '111 222 333',
-                'data_format': 'html'
-            },
-            score_category='irelevant',
-            status=suggestion_models.STATUS_IN_REVIEW,
-            target_type='exploration',
-            target_id=self.EXP_1_ID,
-            target_version_at_submission=0,
-            language_code=self.LANG_1
-        )
-        suggestion_model.update_timestamps()
-        suggestion_model.put()
-        opportunity_model = self.create_model(
-            opportunity_models.ExplorationOpportunitySummaryModel,
-            id=self.EXP_1_ID,
-            topic_id=self.TOPIC_1_ID,
-            chapter_title='irelevant',
-            content_count=1,
-            story_id='irelevant',
-            story_title='irelevant',
-            topic_name='irelevant'
-        )
-        opportunity_model.update_timestamps()
-        opportunity_model.put()
+        mocked_now = datetime.datetime(2025, 1, 7)
+        with self.mock_datetime_utcnow(mocked_now):
+            suggestion_model = self.create_model(
+                suggestion_models.GeneralSuggestionModel,
+                suggestion_type=feconf.SUGGESTION_TYPE_TRANSLATE_CONTENT,
+                author_id=self.VALID_USER_ID_1,
+                change_cmd={
+                    'cmd': exp_domain.CMD_ADD_WRITTEN_TRANSLATION,
+                    'state_name': 'state',
+                    'content_id': 'content_id',
+                    'language_code': 'lang',
+                    'content_html': '111 222 333',
+                    'translation_html': '111 222 333',
+                    'data_format': 'html'
+                },
+                score_category='irelevant',
+                status=suggestion_models.STATUS_IN_REVIEW,
+                target_type='exploration',
+                target_id=self.EXP_1_ID,
+                target_version_at_submission=0,
+                language_code=self.LANG_1
+            )
+            suggestion_model.update_timestamps()
+            suggestion_model.put()
+            opportunity_model = self.create_model(
+                opportunity_models.ExplorationOpportunitySummaryModel,
+                id=self.EXP_1_ID,
+                topic_id=self.TOPIC_1_ID,
+                chapter_title='irelevant',
+                content_count=1,
+                story_id='irelevant',
+                story_title='irelevant',
+                topic_name='irelevant'
+            )
+            opportunity_model.update_timestamps()
+            opportunity_model.put()
 
         self.assert_job_output_is([
             job_run_result.JobRunResult(
@@ -422,41 +431,43 @@ class GenerateContributionStatsJobTests(job_test_utils.JobTestBase):
     def test_creates_translation_stats_models_from_one_accepted_suggestion(
         self
     ) -> None:
-        suggestion_model = self.create_model(
-            suggestion_models.GeneralSuggestionModel,
-            suggestion_type=feconf.SUGGESTION_TYPE_TRANSLATE_CONTENT,
-            author_id=self.VALID_USER_ID_1,
-            change_cmd={
-                'cmd': exp_domain.CMD_ADD_WRITTEN_TRANSLATION,
-                'state_name': 'state',
-                'content_id': 'content_id',
-                'language_code': 'lang',
-                'content_html': '111 222 333',
-                'translation_html': '111 222 333',
-                'data_format': 'unicode'
-            },
-            score_category='irelevant',
-            status=suggestion_models.STATUS_ACCEPTED,
-            target_type='exploration',
-            target_id=self.EXP_1_ID,
-            target_version_at_submission=0,
-            language_code=self.LANG_1,
-            final_reviewer_id='reviewer1'
-        )
-        suggestion_model.update_timestamps()
-        suggestion_model.put()
-        opportunity_model = self.create_model(
-            opportunity_models.ExplorationOpportunitySummaryModel,
-            id=self.EXP_1_ID,
-            topic_id=self.TOPIC_1_ID,
-            chapter_title='irelevant',
-            content_count=1,
-            story_id='irelevant',
-            story_title='irelevant',
-            topic_name='irelevant'
-        )
-        opportunity_model.update_timestamps()
-        opportunity_model.put()
+        mocked_now = datetime.datetime(2025, 1, 7)
+        with self.mock_datetime_utcnow(mocked_now):
+            suggestion_model = self.create_model(
+                suggestion_models.GeneralSuggestionModel,
+                suggestion_type=feconf.SUGGESTION_TYPE_TRANSLATE_CONTENT,
+                author_id=self.VALID_USER_ID_1,
+                change_cmd={
+                    'cmd': exp_domain.CMD_ADD_WRITTEN_TRANSLATION,
+                    'state_name': 'state',
+                    'content_id': 'content_id',
+                    'language_code': 'lang',
+                    'content_html': '111 222 333',
+                    'translation_html': '111 222 333',
+                    'data_format': 'unicode'
+                },
+                score_category='irelevant',
+                status=suggestion_models.STATUS_ACCEPTED,
+                target_type='exploration',
+                target_id=self.EXP_1_ID,
+                target_version_at_submission=0,
+                language_code=self.LANG_1,
+                final_reviewer_id='reviewer1'
+            )
+            suggestion_model.update_timestamps()
+            suggestion_model.put()
+            opportunity_model = self.create_model(
+                opportunity_models.ExplorationOpportunitySummaryModel,
+                id=self.EXP_1_ID,
+                topic_id=self.TOPIC_1_ID,
+                chapter_title='irelevant',
+                content_count=1,
+                story_id='irelevant',
+                story_title='irelevant',
+                topic_name='irelevant'
+            )
+            opportunity_model.update_timestamps()
+            opportunity_model.put()
 
         self.assert_job_output_is([
             job_run_result.JobRunResult(
@@ -525,11 +536,11 @@ class GenerateContributionStatsJobTests(job_test_utils.JobTestBase):
             translation_review_stats_model.accepted_translation_word_count, 3)
         self.assertEqual(
             translation_review_stats_model.first_contribution_date,
-            datetime.datetime.utcnow().date()
+            mocked_now.date()
         )
         self.assertEqual(
             translation_review_stats_model.last_contribution_date,
-            datetime.datetime.utcnow().date()
+            mocked_now.date()
         )
 
     def test_escapes_stats_without_opportunity(
@@ -574,65 +585,67 @@ class GenerateContributionStatsJobTests(job_test_utils.JobTestBase):
     def test_creates_translation_stats_models_from_two_accepted_suggestions(
         self
     ) -> None:
-        first_suggestion_model = self.create_model(
-            suggestion_models.GeneralSuggestionModel,
-            suggestion_type=feconf.SUGGESTION_TYPE_TRANSLATE_CONTENT,
-            author_id=self.VALID_USER_ID_1,
-            change_cmd={
-                'cmd': exp_domain.CMD_ADD_WRITTEN_TRANSLATION,
-                'state_name': 'state',
-                'content_id': 'content_id',
-                'language_code': 'lang',
-                'content_html': '111 222 333',
-                'translation_html': '111 222 333',
-                'data_format': 'unicode'
-            },
-            score_category='irelevant',
-            status=suggestion_models.STATUS_ACCEPTED,
-            target_type='exploration',
-            target_id=self.EXP_1_ID,
-            target_version_at_submission=0,
-            language_code=self.LANG_1,
-            final_reviewer_id=None
-        )
-        first_suggestion_model.update_timestamps()
-        first_suggestion_model.put()
-        opportunity_model = self.create_model(
-            opportunity_models.ExplorationOpportunitySummaryModel,
-            id=self.EXP_1_ID,
-            topic_id=self.TOPIC_1_ID,
-            chapter_title='irelevant',
-            content_count=1,
-            story_id='irelevant',
-            story_title='irelevant',
-            topic_name='irelevant'
-        )
-        opportunity_model.update_timestamps()
-        opportunity_model.put()
+        mocked_now = datetime.datetime(2025, 1, 7)
+        with self.mock_datetime_utcnow(mocked_now):
+            first_suggestion_model = self.create_model(
+                suggestion_models.GeneralSuggestionModel,
+                suggestion_type=feconf.SUGGESTION_TYPE_TRANSLATE_CONTENT,
+                author_id=self.VALID_USER_ID_1,
+                change_cmd={
+                    'cmd': exp_domain.CMD_ADD_WRITTEN_TRANSLATION,
+                    'state_name': 'state',
+                    'content_id': 'content_id',
+                    'language_code': 'lang',
+                    'content_html': '111 222 333',
+                    'translation_html': '111 222 333',
+                    'data_format': 'unicode'
+                },
+                score_category='irelevant',
+                status=suggestion_models.STATUS_ACCEPTED,
+                target_type='exploration',
+                target_id=self.EXP_1_ID,
+                target_version_at_submission=0,
+                language_code=self.LANG_1,
+                final_reviewer_id=None
+            )
+            first_suggestion_model.update_timestamps()
+            first_suggestion_model.put()
+            opportunity_model = self.create_model(
+                opportunity_models.ExplorationOpportunitySummaryModel,
+                id=self.EXP_1_ID,
+                topic_id=self.TOPIC_1_ID,
+                chapter_title='irelevant',
+                content_count=1,
+                story_id='irelevant',
+                story_title='irelevant',
+                topic_name='irelevant'
+            )
+            opportunity_model.update_timestamps()
+            opportunity_model.put()
 
-        second_suggestion_model = self.create_model(
-            suggestion_models.GeneralSuggestionModel,
-            suggestion_type=feconf.SUGGESTION_TYPE_TRANSLATE_CONTENT,
-            author_id=self.VALID_USER_ID_1,
-            change_cmd={
-                'cmd': exp_domain.CMD_ADD_WRITTEN_TRANSLATION,
-                'state_name': 'state',
-                'content_id': 'content_id',
-                'language_code': 'lang',
-                'content_html': '111 222 333',
-                'translation_html': '111 222 333',
-                'data_format': 'unicode'
-            },
-            score_category='irelevant',
-            status=suggestion_models.STATUS_ACCEPTED,
-            target_type='exploration',
-            target_id=self.EXP_1_ID,
-            target_version_at_submission=0,
-            language_code=self.LANG_1,
-            final_reviewer_id=None
-        )
-        second_suggestion_model.update_timestamps()
-        second_suggestion_model.put()
+            second_suggestion_model = self.create_model(
+                suggestion_models.GeneralSuggestionModel,
+                suggestion_type=feconf.SUGGESTION_TYPE_TRANSLATE_CONTENT,
+                author_id=self.VALID_USER_ID_1,
+                change_cmd={
+                    'cmd': exp_domain.CMD_ADD_WRITTEN_TRANSLATION,
+                    'state_name': 'state',
+                    'content_id': 'content_id',
+                    'language_code': 'lang',
+                    'content_html': '111 222 333',
+                    'translation_html': '111 222 333',
+                    'data_format': 'unicode'
+                },
+                score_category='irelevant',
+                status=suggestion_models.STATUS_ACCEPTED,
+                target_type='exploration',
+                target_id=self.EXP_1_ID,
+                target_version_at_submission=0,
+                language_code=self.LANG_1,
+                final_reviewer_id=None
+            )
+            second_suggestion_model.update_timestamps()
+            second_suggestion_model.put()
 
         self.assert_job_output_is([
             job_run_result.JobRunResult(
@@ -702,11 +715,11 @@ class GenerateContributionStatsJobTests(job_test_utils.JobTestBase):
             translation_review_stats_model.accepted_translation_word_count, 6)
         self.assertEqual(
             translation_review_stats_model.first_contribution_date,
-            datetime.datetime.utcnow().date()
+            mocked_now.date()
         )
         self.assertEqual(
             translation_review_stats_model.last_contribution_date,
-            datetime.datetime.utcnow().date()
+            mocked_now.date()
         )
 
     def test_creates_multiple_stats_models_from_multiple_users(
@@ -1061,7 +1074,9 @@ class GenerateContributionStatsJobTests(job_test_utils.JobTestBase):
     def test_creates_question_stats_models_from_one_accepted_suggestion(
         self
     ) -> None:
-        topic_id = self._create_question()
+        mocked_now = datetime.datetime(2025, 1, 7)
+        with self.mock_datetime_utcnow(mocked_now):
+            topic_id = self._create_question()
 
         self.assert_job_output_is([
             job_run_result.JobRunResult(
@@ -1108,11 +1123,11 @@ class GenerateContributionStatsJobTests(job_test_utils.JobTestBase):
         )
         self.assertEqual(
             question_stats_model.first_contribution_date,
-            datetime.datetime.utcnow().date()
+            mocked_now.date()
         )
         self.assertEqual(
             question_stats_model.last_contribution_date,
-            datetime.datetime.utcnow().date()
+            mocked_now.date()
         )
 
         self.assertEqual(
@@ -1131,94 +1146,96 @@ class GenerateContributionStatsJobTests(job_test_utils.JobTestBase):
         )
         self.assertEqual(
             question_review_stats_model.first_contribution_date,
-            datetime.datetime.utcnow().date()
+            mocked_now.date()
         )
         self.assertEqual(
             question_review_stats_model.last_contribution_date,
-            datetime.datetime.utcnow().date()
+            mocked_now.date()
         )
 
     def test_creates_stats_model_from_multiple_suggestions(self) -> None:
-        opportunity_model = self.create_model(
-            opportunity_models.ExplorationOpportunitySummaryModel,
-            id=self.EXP_1_ID,
-            topic_id=self.TOPIC_1_ID,
-            chapter_title='irelevant',
-            content_count=1,
-            story_id='irelevant',
-            story_title='irelevant',
-            topic_name='irelevant'
-        )
-        opportunity_model.update_timestamps()
-        opportunity_model.put()
-        suggestion_1_model = self.create_model(
-            suggestion_models.GeneralSuggestionModel,
-            suggestion_type=feconf.SUGGESTION_TYPE_TRANSLATE_CONTENT,
-            author_id=self.VALID_USER_ID_1,
-            change_cmd={
-                'cmd': exp_domain.CMD_ADD_WRITTEN_TRANSLATION,
-                'state_name': 'state',
-                'content_id': 'content_id',
-                'language_code': 'lang',
-                'content_html': '111 222 333',
-                'translation_html': '111 222 333',
-                'data_format': 'html'
-            },
-            score_category='irelevant',
-            status=suggestion_models.STATUS_IN_REVIEW,
-            target_type='exploration',
-            target_id=self.EXP_1_ID,
-            target_version_at_submission=0,
-            language_code=self.LANG_1,
-            created_on=datetime.datetime.utcnow()
-        )
-        suggestion_1_model.update_timestamps()
-        suggestion_2_model = self.create_model(
-            suggestion_models.GeneralSuggestionModel,
-            suggestion_type=feconf.SUGGESTION_TYPE_TRANSLATE_CONTENT,
-            author_id=self.VALID_USER_ID_1,
-            change_cmd={
-                'cmd': exp_domain.CMD_ADD_WRITTEN_TRANSLATION,
-                'state_name': 'state',
-                'content_id': 'content_id',
-                'language_code': 'lang',
-                'content_html': ['111', '222', '333', '444', '555'],
-                'translation_html': ['111', '222', '333', '444', '555'],
-                'data_format': 'set_of_unicode_string'
-            },
-            score_category='irelevant',
-            status=suggestion_models.STATUS_IN_REVIEW,
-            target_type='exploration',
-            target_id=self.EXP_1_ID,
-            target_version_at_submission=0,
-            language_code=self.LANG_1,
-            created_on=datetime.datetime.utcnow() - datetime.timedelta(days=2)
-        )
-        suggestion_2_model.update_timestamps()
-        suggestion_3_model = self.create_model(
-            suggestion_models.GeneralSuggestionModel,
-            suggestion_type=feconf.SUGGESTION_TYPE_TRANSLATE_CONTENT,
-            author_id=self.VALID_USER_ID_1,
-            change_cmd={
-                'cmd': exp_domain.CMD_ADD_WRITTEN_TRANSLATION,
-                'state_name': 'state',
-                'content_id': 'content_id',
-                'language_code': 'lang',
-                'content_html': ['111', '222', '333', '444', '555'],
-                'translation_html': ['111', '222', '333', '444', '555'],
-                'data_format': 'set_of_unicode_string'
-            },
-            score_category='irelevant',
-            status=suggestion_models.STATUS_IN_REVIEW,
-            target_type='exploration',
-            target_id=self.EXP_1_ID,
-            target_version_at_submission=0,
-            language_code=self.LANG_1,
-            created_on=datetime.datetime.utcnow() - datetime.timedelta(days=1)
-        )
-        suggestion_3_model.update_timestamps()
-        suggestion_models.GeneralSuggestionModel.put_multi([
-            suggestion_1_model, suggestion_2_model, suggestion_3_model])
+        mocked_now = datetime.datetime(2025, 1, 7)
+        with self.mock_datetime_utcnow(mocked_now):
+            opportunity_model = self.create_model(
+                opportunity_models.ExplorationOpportunitySummaryModel,
+                id=self.EXP_1_ID,
+                topic_id=self.TOPIC_1_ID,
+                chapter_title='irelevant',
+                content_count=1,
+                story_id='irelevant',
+                story_title='irelevant',
+                topic_name='irelevant'
+            )
+            opportunity_model.update_timestamps()
+            opportunity_model.put()
+            suggestion_1_model = self.create_model(
+                suggestion_models.GeneralSuggestionModel,
+                suggestion_type=feconf.SUGGESTION_TYPE_TRANSLATE_CONTENT,
+                author_id=self.VALID_USER_ID_1,
+                change_cmd={
+                    'cmd': exp_domain.CMD_ADD_WRITTEN_TRANSLATION,
+                    'state_name': 'state',
+                    'content_id': 'content_id',
+                    'language_code': 'lang',
+                    'content_html': '111 222 333',
+                    'translation_html': '111 222 333',
+                    'data_format': 'html'
+                },
+                score_category='irelevant',
+                status=suggestion_models.STATUS_IN_REVIEW,
+                target_type='exploration',
+                target_id=self.EXP_1_ID,
+                target_version_at_submission=0,
+                language_code=self.LANG_1,
+                created_on=datetime.datetime.utcnow()
+            )
+            suggestion_1_model.update_timestamps()
+            suggestion_2_model = self.create_model(
+                suggestion_models.GeneralSuggestionModel,
+                suggestion_type=feconf.SUGGESTION_TYPE_TRANSLATE_CONTENT,
+                author_id=self.VALID_USER_ID_1,
+                change_cmd={
+                    'cmd': exp_domain.CMD_ADD_WRITTEN_TRANSLATION,
+                    'state_name': 'state',
+                    'content_id': 'content_id',
+                    'language_code': 'lang',
+                    'content_html': ['111', '222', '333', '444', '555'],
+                    'translation_html': ['111', '222', '333', '444', '555'],
+                    'data_format': 'set_of_unicode_string'
+                },
+                score_category='irelevant',
+                status=suggestion_models.STATUS_IN_REVIEW,
+                target_type='exploration',
+                target_id=self.EXP_1_ID,
+                target_version_at_submission=0,
+                language_code=self.LANG_1,
+                created_on=datetime.datetime.utcnow() - datetime.timedelta(days=2)
+            )
+            suggestion_2_model.update_timestamps()
+            suggestion_3_model = self.create_model(
+                suggestion_models.GeneralSuggestionModel,
+                suggestion_type=feconf.SUGGESTION_TYPE_TRANSLATE_CONTENT,
+                author_id=self.VALID_USER_ID_1,
+                change_cmd={
+                    'cmd': exp_domain.CMD_ADD_WRITTEN_TRANSLATION,
+                    'state_name': 'state',
+                    'content_id': 'content_id',
+                    'language_code': 'lang',
+                    'content_html': ['111', '222', '333', '444', '555'],
+                    'translation_html': ['111', '222', '333', '444', '555'],
+                    'data_format': 'set_of_unicode_string'
+                },
+                score_category='irelevant',
+                status=suggestion_models.STATUS_IN_REVIEW,
+                target_type='exploration',
+                target_id=self.EXP_1_ID,
+                target_version_at_submission=0,
+                language_code=self.LANG_1,
+                created_on=datetime.datetime.utcnow() - datetime.timedelta(days=1)
+            )
+            suggestion_3_model.update_timestamps()
+            suggestion_models.GeneralSuggestionModel.put_multi([
+                suggestion_1_model, suggestion_2_model, suggestion_3_model])
 
         self.assert_job_output_is([
             job_run_result.JobRunResult(

